std-bits/base/src/main/java/org/enso/base/parser/FormatDetectingNumberParser.java

```diff
@@ -0,0 +1,309 @@
+ *   <li>BRACKETS + SYMBOL + NUMBER + BRACKET_CLOSE
+ *   <li>BRACKETS + NUMBER + SYMBOL + BRACKET_CLOSE
+ * </ul>
+ */
+public class FormatDetectingNumberParser {
+  public interface NumberParseResult {}
+
+  public interface NumberParseResultSuccess extends NumberParseResult {
+    NumberParseResultSuccess negate();
+
+    NumberParseResultSuccess withSymbol(String symbol);
+  }
+
+  public record NumberParseLong(long number, String symbol) implements NumberParseResultSuccess {
+    @Override
+    public NumberParseResultSuccess negate() {
+      return new NumberParseLong(-number, symbol);
+    }
+
+    @Override
+    public NumberParseResultSuccess withSymbol(String symbol) {
+      return new NumberParseLong(number, symbol);
+    }
+  }
+
+  public record NumberParseDouble(double number, String symbol)
+      implements NumberParseResultSuccess {
+    @Override
+    public NumberParseResultSuccess negate() {
+      return new NumberParseDouble(-number, symbol);
+    }
+
+    @Override
+    public NumberParseResultSuccess withSymbol(String symbol) {
+      return new NumberParseDouble(number, symbol);
+    }
+  }
+
+  public record NumberParseFailure(String message) implements NumberParseResult {}
+
+  private final boolean allowSymbol;
+  private final boolean allowLeadingZeroes;
+  private final boolean allowLeadingTrailingWhitespace;
+  private NegativeSign negativeSign;
+  private NumberWithSeparators numberWithSeparators;
+
+  public FormatDetectingNumberParser() {
+    this(true, true, true, NegativeSign.UNKNOWN, NumberWithSeparators.UNKNOWN);
+  }
+
+  public FormatDetectingNumberParser(
+      boolean allowSymbol,
+      boolean allowLeadingZeroes,
+      boolean allowLeadingTrailingWhitespace,
+      NegativeSign negativeSign,
+      NumberWithSeparators numberWithSeparators) {
+    this.allowSymbol = allowSymbol;
+    this.allowLeadingZeroes = allowLeadingZeroes;
+    this.allowLeadingTrailingWhitespace = allowLeadingTrailingWhitespace;
+    this.negativeSign = negativeSign;
+    this.numberWithSeparators = numberWithSeparators;
+  }
+
+  public NegativeSign negativeSign() {
+    return negativeSign;
+  }
+
+  public void setNegativeSign(NegativeSign newNegativeSign) {
+    if (negativeSign != NegativeSign.UNKNOWN) {
+      throw new IllegalStateException("Negative Sign Already Set.");
+    }
+    negativeSign = newNegativeSign;
+  }
+
+  public NumberWithSeparators numberWithSeparators() {
+    return numberWithSeparators;
+  }
+
+  /**
+   * Parse a string into a number.
+   *
+   * @param value the string to parse.
+   * @param integer whether to parse a Long or a Double.
+   * @return the parsed number, or a failure if the parse was unsuccessful.
+   */
+  public NumberParseResult parse(CharSequence value, boolean integer) {
+    // State
+    boolean lastWasWhitespace = false;
+    boolean encounteredContent = false;
+    boolean encounteredSign = false;
+    boolean isNegative = false;
```

Comment:
Looking at usages of this variable, `needsNegate` or `shouldNegate` seem like a better option.

Once we call `negate`, changing `isNegative` to `false` is IMO misleading - the number **is still negative**, it just has already been negated so does not need the `negate` call anymore. A different name would make the intent clearer.